### PR TITLE
LPS-164897 Notify subscribed users on Revision Date

### PR DESCRIPTION
**Pull request by marcogalluzzi**
Now emails and web notifications are sent to Owners and Subscribers of KB Articles when the Review date expires.

Also, the configuration of the email subject and body for an expired review has been added.

This FF must be enabled to work with Review dates: **feature.flag.LPS-165476**

Related issue: https://issues.liferay.com/browse/LPS-164897

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 1]
```diff
@@ -1372,6 +1377,8 @@ public KBArticle updateStatus(
 			return kbArticle;
 		}
 
+		String action = "add";
```
Comment (adolfopa) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1380]: Where possible, can you use already existing constants in the `Constants` class (portal-kernel)? There are at least two that could be used here (ADD and UPDATE).

**Review by adolfopa — CHANGES_REQUESTED** [review]

**Comment by AliciaGarciaGarcia** [comment]
I'll resume this task!

Thanks!

**Comment by AliciaGarciaGarcia** [comment]
@marcogalluzzi I had no time to resume this task, sorry

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 1 above)
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java]: I've followed your advice, thanks

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by marcogalluzzi** [comment]
ci:stop

**Comment by marcogalluzzi** [comment]
ci:test:stop

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by marcogalluzzi** [comment]
ci:stop

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by marcogalluzzi** [comment]
ci:stop

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 2]
```diff
@@ -344,7 +347,18 @@ public void addTempAttachment(
...
 	public void checkKBArticles() throws PortalException {
-		_checkKBArticles(new Date());
+		Date date = new Date();
+
+		_checkKBArticlesByExpirationDate(date);
+
+		if (_previousCheckDate == null) {
+			_previousCheckDate = new Date(
+				date.getTime() - _getKBArticleCheckInterval());
+		}
```
Comment (adolfopa) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:357]: Can we remove `previousCheckDate`? For what I've seen, it is only used in the query, but I don't think it is strictly necessary. It seems wrong to keep mutable state in a singleton service.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 3]
```diff
@@ -1956,73 +2062,107 @@ private void _notifySubscribers(
...
+			serviceContext.setPortalURL(
+				company.getPortalURL(kbArticle.getGroupId()));
+			serviceContext.setPortletId(KBPortletKeys.KNOWLEDGE_BASE_ADMIN);
+			serviceContext.setScopeGroupId(kbArticle.getGroupId());
+
+			_notify(
+				new HashSet<>(
+					Arrays.asList(
+						_NOTIFICATION_RECEIVER_OWNER,
+						_NOTIFICATION_RECEIVER_SUBSCRIBER)),
```
Comment (adolfopa) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:2161]: Can you use `SetUtil.fromArray` instead?

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 2 above)
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:357]: After internal discussion, it seems not an easy thing to solve. We can think of a better solution later.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 3 above)
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:2161]: Yes thanks. I didn't know about it.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 4]
```diff
@@ -1764,6 +1795,56 @@ private List<KBArticle> _getKBArticlesByCompanyIdAndExpirationDate(
...
+				).and(
+					KBArticleTable.INSTANCE.reviewDate.lte(reviewDateLTE)
+				)
+			));
+	}
+
+	private String _getKBArticleURL(KBArticle kbArticle)
+		throws PortalException {
+
+		String controlPanelFullURL = _portal.getControlPanelFullURL(
```
Comment (AliciaGarciaGarcia) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1820]: have you tried to use 
```
return PortletURLBuilder.create(
			_portal.getControlPanelPortletURL(
				httpServletRequest, 
				...
```

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/notifications/DLUserNotificationHandler.java [hunk 5]
```diff
@@ -71,7 +75,9 @@ else if (notificationType ==
 		else if (notificationType ==
 					UserNotificationDefinition.NOTIFICATION_TYPE_REVIEW_ENTRY) {
 
-			message = "x-review-a-x";
+			return _language.format(
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/notifications/DLUserNotificationHandler.java:78]: The review message `x-review-a-x` doesn't make sense so I created a new message `x-needs-review` in KB, and for the sake of improving I changed the message also in DL.

File: modules/apps/knowledge-base/knowledge-base-api/src/main/resources/com/liferay/knowledge/base/configuration/packageinfo [hunk 6]
```diff
@@ -1 +1 @@
-version 3.0.0
\ No newline at end of file
+version 4.0.0
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-api/src/main/resources/com/liferay/knowledge/base/configuration/packageinfo:1]: Yes thanks!

File: modules/apps/knowledge-base/knowledge-base-api/src/main/resources/com/liferay/knowledge/base/dependencies/email_kb_article_review_body.tmpl [hunk 7]
```diff
@@ -0,0 +1,13 @@
+Dear [$TO_NAME$],<br /><br />
+
+This is an autogenerated email for the [$PORTLET_NAME$] portlet from site [$SITE_NAME$].<br /><br />
+
+Please review the article in category [$CATEGORY_TITLE$] with title [$ARTICLE_TITLE$] and version [$ARTICLE_VERSION$].<br /><br />
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-api/src/main/resources/com/liferay/knowledge/base/dependencies/email_kb_article_review_body.tmpl:5]: I can put the title with bold or italic. About the text, I've copied it from another template. If you have a better proposal fort text I can use it.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 8]
```diff
@@ -1910,30 +2010,36 @@ private String _normalizeUrlTitle(String urlTitle) {
...
 
-		if (serviceContext.isCommandAdd() &&
+		if (Constants.ADD.equals(action) &&
 			!kbGroupServiceConfiguration.emailKBArticleAddedEnabled()) {
 
 			return;
 		}
 
-		if (serviceContext.isCommandUpdate() &&
+		if (Constants.UPDATE.equals(action) &&
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:2031]: I'll change it. Thanks.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 4 above)
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1820]: Before my changes, there method `_getKBArticleURL` was using the `PortletURLBuilder` as you suggest. However, this builder needs an `HTTPServletRequest`, which doesn't exist in the new review action.

**Comment by marcogalluzzi** [comment]
I've added requested changes in commit "made little improvements"

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by AliciaGarciaGarcia** [comment]
Hi @marcogalluzzi 

maybe [functional-tomcat90-mysql57-chrome100-jdk8/2/9](https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/130734//consoleText)
[LocalFile.DMNotifications#CanViewEmailNotificationForNewDocumentInFolder](https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/130734//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_DMNotifications_CanViewEmailNotificationForNewDocumentInFolder_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/LocalFile.DMNotifications_CanViewEmailNotificationForNewDocumentInFolder/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/LocalFile.DMNotifications_CanViewEmailNotificationForNewDocumentInFolder/summary.html.gz) - [Console Output](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/jenkins-console.txt.gz) - [Failed 0 of last 25](https://testray.liferay.com/home/-/testray/case_results/2166329513/history) is relevant for this PR, can you check it?

thanks!

**Comment by marcogalluzzi** [comment]
> Hi @marcogalluzzi
> 
> maybe [functional-tomcat90-mysql57-chrome100-jdk8/2/9](https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/130734//consoleText) [LocalFile.DMNotifications#CanViewEmailNotificationForNewDocumentInFolder](https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/130734//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_DMNotifications_CanViewEmailNotificationForNewDocumentInFolder_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/LocalFile.DMNotifications_CanViewEmailNotificationForNewDocumentInFolder/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/LocalFile.DMNotifications_CanViewEmailNotificationForNewDocumentInFolder/summary.html.gz) - [Console Output](https://testray.liferay.com/reports/production/logs/2023-01/test-1-35/test-portal-acceptance-pullrequest(master)/3384/functional-tomcat90-mysql57-chrome100-jdk8/2/9/jenkins-console.txt.gz) - [Failed 0 of last 25](https://testray.liferay.com/home/-/testray/case_results/2166329513/history) is relevant for this PR since the message change, can you check it?
> 
> thanks!

The error doesn't seem to be related: the POSHI report says it's not finding the "Edit" menu item, when clicking on a DM document

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:test:relevant

**Review by AliciaGarciaGarcia — APPROVED** [review]

**Comment by AliciaGarciaGarcia** [comment]
ci:reevaluate:1267401_7174

**Comment by marcogalluzzi** [comment]
ci:reevaluate:1267401_7174

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Comment by AliciaGarciaGarcia** [comment]
ci:forward

**Pull request closed without merging**